conda_build/utils.py

```diff
@@ -718,7 +721,128 @@ def __exit__(self, et, ev, tb):
         # implicit return of None => don't swallow exceptions
 
 
+def get_installed_packages(path):
+    '''
+    Scan all json files in 'path' and return a dictionary with their contents.
+    Files are assumed to be in 'index.json' format.
+    '''
+    installed = dict()
+    for filename in glob(os.path.join(path, 'conda-meta', '*.json')):
+        with open(filename) as file:
+            data = json.load(file)
+            installed[data['name']] = data
+    return installed
+
+
+def _convert_lists_to_sets(_dict):
+    for k, v in _dict.items():
+        if hasattr(v, 'keys'):
+            _dict[k] = HashableDict(_convert_lists_to_sets(v))
+        elif hasattr(v, '__iter__'):
+            _dict[k] = sorted(list(set(v)))
+    return _dict
+
+
+class HashableDict(dict):
+    """use hashable frozen dictionaries for resources and resource types so that they can be in sets
+    """
+    def __init__(self, *args, **kwargs):
+        super(HashableDict, self).__init__(*args, **kwargs)
+        self = _convert_lists_to_sets(self)
+
+    def __hash__(self):
+        return hash(json.dumps(self, sort_keys=True))
```

Comment:
@msarahan 4 hex characters will contain 16 bits of the 160 bit sha1 hash (4 bits per hex character).  Fortunately, we should be able to estimate if its good enough by calculating the probability of a hash collision.  If we have 20 variants, the probability of a collision is ~2.9%.   If we do those same 20 variants with 20 bits (5 hex chars), the probability of a collision drops to ~.02%.  How many variants of the same package do we want to be able to handle?

Method described here: http://preshing.com/20110504/hash-collision-probabilities/
Convenient calculator here :) http://everydayinternetstuff.com/2015/04/hash-collision-probability-calculator/